### PR TITLE
Run Apply's structure gate at approve time (approve == apply, #1416)

### LINKED ISSUE #1416: Approve transition skips proposal structure validation (zero-op proposal approvable; Apply then 400s)
## Seam

`AutomationProposalService.ApproveProposalAsync` (`backend/src/Taskdeck.Application/Services/AutomationProposalService.cs` ~223-244).

## Gap

Approve does not validate proposal structure: a zero-operation `PendingReview` proposal can be approved (status → `Approved`) and only fails at Apply/Execute time with 400 "Proposal must contain at least one operation". The #1395/#1413 slices made the DIFF path run Apply's gates; the APPROVE transition still runs none of them, so the reviewer's approval action itself can succeed on a proposal that Apply must reject — the last remaining "user commits to something the executor will refuse" step in this trust class.

PR #1414 (issue #1397) narrows this on the frontend (invalid-state presentation + approve-rail operation-count gate), but the server-side transition remains permissive for any client (MCP, CLI, raw HTTP).

## Suggested direction

Run the structure gate (`ProposalOperationStructureValidator` — the zero-op check at minimum) at approve time, mirroring the diff-path pattern from #1376/#1398, with the same 400 `ValidationError` shape. Decide deliberately whether the expiry gate also belongs at approve time (an expired proposal currently 409s/400s elsewhere — keep semantics consistent with the executor's terminal-status short-circuit).

## Provenance

- PR #1414 round-1 adversarial review (product-contract lens, finding 3), coordinator adjudication 2026-07-17 overnight run — real, backend-scoped, out of the frontend PR's scope; tracked per review policy.
- Trust-class siblings: #1370 → #1374, #1376 → #1395, #1398 → #1413, #1415 (MCP resource surface).

### PR DESCRIPTION
**Pull request by Chris0Jeky**
## Summary

Closes #1416.

The Approve transition skipped proposal validation: a zero-operation `PendingReview` proposal could be approved (status -> `Approved`) and only failed later at Apply with 400 "Proposal must contain at least one operation" — and a structurally valid proposal whose requester lost board access (or whose board/requester was deleted, or whose revised operations violate the operation contract) likewise approved 200 and only failed at Apply. This was the last "user commits to something the executor will refuse" step in the preview == apply trust class (siblings #1370 -> #1374, #1376 -> #1395, #1398 -> #1413).

`AutomationProposalService.ApproveProposalAsync` now runs Apply's structure AND permission/contract gates before committing the transition (permission gate added in the lens-2 review fix round, commit `48ccf0ef`).

## Gates enforced at approve time

**1. Structure gate (400 `ValidationError`)** — mirrors the apply path exactly. Apply validates the EFFECTIVE operation set (latest saved revision, else the original operations) via `AutomationPolicyEngine.ValidateOperationStructure` (`AutomationPolicyEngine.cs:92-108` -> `ProposalOperationStructureValidator.Validate`), and `GetProposalDiffAsync` mirrors the same revision-aware materialization. Approve resolves the same effective operations (revision-aware, mirroring `AutomationExecutorService.MaterializeEffectiveProposalAsync`, `AutomationExecutorService.cs:241-261`) and runs the identical validator, returning the same 400 `ValidationError` "Proposal must contain at least one operation" the diff/apply paths return.

- Revision-aware on purpose: a revision is structure-validated at save time (`ProposalRevisionService.cs:46-53`; `ProposalRevisionPayload.TryParseOperations` rejects an empty operations array), so validating only the original operations would be wrong in both directions — falsely rejecting an originally-empty-but-validly-revised proposal Apply accepts, and falsely approving an originally-valid-but-revised-to-empty proposal Apply refuses. Both directions are test-pinned.

**2. Expiry gate (409 `InvalidOperation`)** — already enforced, preserved. The domain transition is the prior art: `AutomationProposal.Approve` throws `DomainException(ErrorCodes.InvalidOperation, "Cannot approve expired proposal")` (`AutomationProposal.cs:96-97`), mapping to 409 (`ResultExtensions.cs:24`). This is the established approve-time expiry contract — a real 409 — and it deliberately does NOT adopt the diff/preview path's 400 "Proposal has expired" read-parity shape, because approving is a state transition and 409 conflict is the correct semantics for refusing to advance an expired proposal.

**3. Permission + operation-contract gate (400/403/404)** — the same `_policyEngine.ValidatePermissionsAsync(requesterId, boardId, effectiveOperations, ct)` call Apply runs (`AutomationExecutorService.cs:138-152`) and the diff path runs post-#1398/#1413: requester exists -> 404, board exists -> 404, requester board access -> 403, then `ProposalOperationContractValidator` -> 400/403/404. Error code and message propagate unchanged, so approve rejects exactly what Apply would reject.

**Gate ordering** mirrors the diff/apply sequence exactly: structure -> expiry -> permissions. The permission gate is skipped for an expired proposal so the domain guard's 409 owns expiry (an expired proposal with revoked access reports expiry, never Forbidden — the #1413 LOW-4 ordering pin adapted to approve's 409 semantics, test-pinned). A zero-op AND expired proposal reports the 400 structure error first, as on diff and apply.

## Deliberately NOT changed

- Reject / Dismiss / Defer / Mark* transitions and every other lifecycle semantic.
- Terminal-status short-circuits: all gates run only for `PendingReview` proposals; any other status keeps the domain 409 "Cannot approve proposal in status X".
- The diff and apply paths (already carry these gates).
- No frontend changes — PR #1414 (issue #1397) handles the frontend presentation separately.
- Approve/reject/get response DTO shape (returns original operations even when a revision exists) — pre-existing, tracked as #1424.

## Verification

- `dotnet build backend/Taskdeck.sln -c Release -m:1` — clean, 0 errors.
- `dotnet test ... Taskdeck.Application.Tests --filter "FullyQualifiedName~AutomationProposalService"` — 86 passed / 0 failed. Approve coverage now pins: zero-op -> 400; expired -> 409; revoked access -> 403 and deleted board -> 404 (each dual-asserted identical to the apply-side engine result); contract-violating revision -> 400; expired+revoked -> 409 (ordering pin); malformed revision payload -> 400; revised-to-empty -> 400; originally-empty-but-validly-revised -> approves; valid proposal -> approves.
- `dotnet test ... Taskdeck.Api.Tests` targeted: ProposalErrorContractTests + AutomationProposalsApiTests — 48 passed (incl. endpoint-level `ApproveProposal_ZeroOperationProposal_Returns400WithErrorContract`); approve-touching regression set (McpTools, ProposalApprovalRace, CaptureToBoardGoldenPath, CrossUserDataIsolation, CardsApi, CaptureApi, TranscriptTriageLlmGoldenPath, LiveBrowserRegressionApi, AuthzRegressionMatrix) — 134 passed; ConcurrencyRaceConditionStressTests — 30 passed.

## Review trail

- Round 1 (self, adversarial): 0 C / 0 H / 0 M / 2 LOW — LOW-1 dismissed (adjudicated SOUND by lens 2), LOW-2 tracked as #1424.
- Round 2 (lens-2 independent, coordinator-adjudicated): MERGE-SAFE, 1 MEDIUM (permission gate — fixed in `48ccf0ef`) + LOW-A (malformed-revision pin) + LOW-B (revised-to-empty pin), all fixed with evidence comment on the PR.

### CODE REVIEW

**Comment by Chris0Jeky** [comment]
## Adversarial Code Review

Reviewed the full diff with a security / preview==apply / GP-06 / HTTP-semantics lens, plus test adequacy.

### CRITICAL
- None.

### HIGH
- None. GP-06 is honoured — the change only makes the approve transition *stricter* (adds a gate); it never lets an agent approve or mutate a board. No auth/cross-user surface touched. Apply still re-validates transactionally at execute time, so nothing here weakens the executor's guarantees.

### MEDIUM
- None. Considered but rejected as in-scope-correct: **approve does not run the permission gate** (`ValidatePermissionsAsync`) that the diff path added in #1398/#1413. Deliberate and safe: (1) #1416 is scoped to structure (+ expiry); (2) permission state is inherently racy between approve and apply, so an approve-time check gives no real guarantee; (3) the executor re-runs `ValidatePermissionsAsync` transactionally at execute, so a requester who lost board access is still blocked at apply — no data exposure. The preview path already surfaces the permission mismatch to the reviewer.

### LOW
- **Gate ordering vs. empty-decider message** (immaterial): the structure gate runs before `proposal.Approve()`, whose first guard is `decidedByUserId == Guid.Empty` -> 400. So a zero-op proposal submitted with an empty decider now reports "Proposal must contain at least one operation" instead of "DecidedByUserId cannot be empty". Both are 400 `ValidationError`, and `decidedByUserId` is claims-derived at the endpoint (`AuthenticatedControllerBase.TryGetCurrentUserId`) so it is never empty in practice — reachable only via direct service calls. No fix warranted; adding a redundant service-layer empty check would duplicate a domain guard.
- **Approve response returns original, not effective, operations** (pre-existing, out of scope): `ApproveProposalAsync` returns `MapToDto(proposal)` built from `proposal.Operations`, so approving a proposal that has a saved revision echoes the original operations even though the gate now validates — and Apply will run — the revised set. This predates this PR (it is true of get/reject/approve alike) and the diff endpoint is the canonical place to see effective operations. Seeding a tracking issue rather than expanding this slice.

### Bot Comments Addressed
- `gemini-code-assist[bot]`: only the "consumer version sunset / review activity ceased" notice — no actionable finding.

### Summary
0 CRITICAL / 0 HIGH / 0 MEDIUM / 2 LOW (1 immaterial-no-fix, 1 pre-existing-tracked). Not merge-blocking. Structure gate mirrors Apply's revision-aware `ValidateOperationStructure` (400 `ValidationError`); expiry preserved as the domain transition's 409 `InvalidOperation`; terminal-status short-circuits untouched (gated to `PendingReview`). Coordinator gates the merge.

**Comment by Chris0Jeky** [comment]
## Adversarial Review — Fixes Applied

No code changes were required: the review found 0 CRITICAL / 0 HIGH / 0 MEDIUM defects. The two LOW items are dispositioned below (nothing silently dropped, per zero-skip policy).

| Finding | Severity | Disposition |
|---------|----------|-------------|
| Permission gate not run at approve time | (considered) MEDIUM | Not a defect — deliberate scope + apply re-validates permissions transactionally at execute; preview already surfaces the mismatch. No action. |
| Gate ordering vs. empty-decider message | LOW | No fix — both paths return 400 `ValidationError`; `decidedByUserId` is claims-derived at the endpoint so an empty value is unreachable in practice. A redundant service-layer check would duplicate a domain guard. |
| Approve/get/reject response returns original (not effective/revised) operations | LOW | Pre-existing, out of scope — tracked as #1424. |

### Verification (unchanged since PR open, re-confirmed)
- Build: `dotnet build backend/Taskdeck.sln -c Release -m:1` — clean, 0 errors.
- `Taskdeck.Application.Tests --filter FullyQualifiedName~AutomationProposalService` — 80 passed / 0 failed (5 new/updated approve tests).
- `Taskdeck.Api.Tests --filter FullyQualifiedName~ProposalErrorContractTests` — 8 passed (incl. new zero-op approve 400 endpoint test).
- Regression: `AutomationProposalsApiTests` (40) + `McpToolsTests`/`ProposalApprovalRaceTests`/`CaptureToBoardGoldenPathIntegrationTests` (40) — all passed.

All findings addressed. CI status: pending checks below.

**Comment by Chris0Jeky** [comment]
## Lens-2 Review — Fixes Applied (batched round)

Lens-2 independent review returned MERGE-SAFE with 1 MEDIUM + 2 LOW; coordinator adjudicated one batched fix round. All findings addressed, zero-skip.

| Finding | Severity | Fix Commit | Tests | Verified |
|---------|----------|-----------|-------|----------|
| Approve mirrored only the structure gate; apply AND diff (#1413) also run `ValidatePermissionsAsync` (requester-exists 404, board-exists 404, board-access 403, operation-contract 400/403/404), so a structurally-valid proposal whose requester lost access (or board/requester deleted, or revised ops fail contract) still approved 200 and only failed at Apply | MEDIUM | `48ccf0ef` (impl) + `684cae2e` (tests) + `629f0569` (fixture) | `ApproveProposalAsync_ShouldRejectRevokedBoardAccess_MatchingApplyPermissionGate` (403, dual-asserted identical to the engine's apply-side result), `ApproveProposalAsync_ShouldRejectDeletedBoard_MatchingApplyPermissionGate` (404, dual-asserted), `ApproveProposalAsync_ShouldRejectContractViolation_OnRevisedOperations` (400, exact contract message), `ApproveProposalAsync_ShouldReportExpiry_NotForbidden_WhenExpiredAndAccessRevoked` (ordering pin: expired + revoked reports the 409 expiry, never Forbidden — mirrors the #1413 LOW-4 pin adapted to approve's 409 semantics) | pass |
| LOW-A: the defensive malformed-revision branch (`ResolveEffectiveOperationsAsync` parse failure) was unpinned on approve | LOW | `684cae2e` | `ApproveProposalAsync_ShouldReturnValidationError_WhenSavedRevisionPayloadIsInvalid` (RevisedPayload `{not valid json` → 400 "RevisedPayload must be valid JSON") | pass |
| LOW-B: inverse revision-aware direction unpinned (originally-valid proposal revised to empty) | LOW | `684cae2e` | `ApproveProposalAsync_ShouldRejectProposalRevisedToEmpty_MatchingApplyGate` (originally-valid + `{"operations":[]}` revision → 400 "RevisedPayload operations must contain at least one operation", transition refused) | pass |
| LOW-1 from round 1 (gate-order vs empty-decider message) | LOW | — | Dismissal adjudicated by lens 2 as SOUND — recorded as adjudicated-invalid, no action | closed |

### Implementation notes (MEDIUM)
- `_policyEngine.ValidatePermissionsAsync(proposal.RequestedByUserId, proposal.BoardId, effectiveOperations, ct)` runs after the structure gate on the same revision-aware effective operation set, propagating the engine's ErrorCode/ErrorMessage unchanged — the identical call and mapping `GetProposalDiffAsync` uses (#1398/#1413) and `AutomationExecutorService.ExecuteProposalAsync` re-runs at apply.
- Ordering mirrors diff/apply exactly (structure → expiry → permissions): the permission gate is skipped when `proposal.IsExpired`, so the domain guard's 409 owns expiry — pinned by the ShouldReportExpiry_NotForbidden test above.
- PendingReview-only, same as the structure gate; terminal-status 409 short-circuits untouched.
- The `ApproveProposalAsync` comment block now states the full contract: structure 400 + expiry 409 (domain guard) + permissions/contract 400/403/404, with the ordering rationale.
- Test collateral: the happy-path/revision-valid/concurrency fixtures now use in-scope board-update operations (and the edge-case fixture gained passing requester/board/access mocks) so they clear the new gate and still exercise their original seams.

### Verification (fix round)
- Build: `dotnet build backend/Taskdeck.sln -c Release -m:1` — clean, 0 errors.
- `Taskdeck.Application.Tests --filter FullyQualifiedName~AutomationProposalService` — **86 passed / 0 failed** (was 80; +6 new). All 15 ApproveProposalAsync tests listed and passing.
- `Taskdeck.Api.Tests --filter ProposalErrorContractTests|AutomationProposalsApiTests` — **48 passed / 0 failed**.
- `Taskdeck.Api.Tests` regression set (McpTools, ProposalApprovalRace, CaptureToBoardGoldenPath, CrossUserDataIsolation, CardsApi, CaptureApi, TranscriptTriageLlmGoldenPath, LiveBrowserRegressionApi, AuthzRegressionMatrix) — **134 passed / 0 failed**.
- `Taskdeck.Api.Tests --filter ConcurrencyRaceConditionStressTests` — **30 passed / 0 failed**.

New head: `629f0569`. CI: pending on this push; previous head was green on all gates.

**Comment by Chris0Jeky** [comment]
## Gate Round — Full-Suite Failure: Classification and Fixes

The coordinator's full-suite gate on `629f0569` failed with 16 failures, all in `ProposalRevisionApiTests` (a file the previous targeted set missed — noted; it is the revision-flow pin family and now belongs in the targeted set for any approve/apply/diff gate change). Everything else was green (7357 passed).

### Failure shape

Every failing test followed the pre-#1416 contract: save a deliberately invalid revision on a PendingReview proposal -> assert preview rejects -> **approve expecting 200** -> assert apply rejects. The new approve-time gates now reject at the approve step.

### Classification (discriminator: would APPLY accept the proposal in the state the test approves it?)

| Test | Apply verdict at approve time | Classification | Action |
|------|------------------------------|----------------|--------|
| `RevisedDueDate_InvalidFormat_*` (2 theory cases) | REFUSE (400, invalid dueDate format) | Deliberate new semantics | Test updated: approve -> 400 ValidationError; execute -> 409 (never approved); renamed `...AndApprove` |
| `RevisedUpdate_WithOnlyFalseClearDueDate` | REFUSE (400, no-op update contract) | Deliberate | Same update + rename |
| `RevisedLabelOperation_WithoutCardId` | REFUSE (400, missing cardId) | Deliberate | Same update + rename |
| `RevisedUpdate_WithEmptyLabelId` | REFUSE (400, empty label identity) | Deliberate | Same update + rename |
| `RevisedLabelOperation_WithUnregisteredAlias_*` (4 theory cases) | REFUSE (400, unregistered alias) | Deliberate | Same update + rename |
| `RevisedCreate_WithoutRequiredTitle` | REFUSE (400, missing create title) | Deliberate | Same update + rename |
| `RevisedUpdate_CannotRedirectAuthorizedTargetToAnotherUsersCard` | REFUSE (400, targetId/cardId mismatch) | Deliberate | approve -> 400; execute -> 409 |
| `RevisedUpdate_CannotTargetAnotherUsersCardWithMatchingIdentities` | REFUSE (403, cross-user card scope) | Deliberate | approve -> 403 Forbidden; execute -> 409 |
| `RevisedCreate_CannotRedirectBoardAndColumnToAnotherUsersBoard` | REFUSE (403, cross-board scope) | Deliberate | approve -> 403; execute -> 409 |
| `RevisedCreate_CannotUseAnotherUsersColumnWithinAuthorizedBoardParameters` | REFUSE (403, cross-board column) | Deliberate | approve -> 403; execute -> 409 |
| `RevisedLabelName_WithAmbiguousBoardMatches_*` (2 theory cases) | REFUSE (400, ambiguity already present) | Deliberate — but the apply-side seam is time-dependent and still reachable | **Restructured**: approve while the label name is unique (valid -> 200, matching apply), THEN create the case-insensitive duplicate, then assert preview AND apply reject the approved proposal (400) — pinning the now-only path to an approved-but-unappliable proposal. Added `RevisedLabelName_AlreadyAmbiguousAtApproveTime_ShouldFailApprove` (2 theory cases) to pin the static direction (approve -> 400, execute -> 409) |

**Zero apply-would-accept cases -> zero gate bugs -> no drift.** Parity evidence from the failure data itself: in every test, approve rejected with the SAME status the test's own diff assertion expects (400 where diff asserts ValidationError, 403 where diff asserts Forbidden) — approve, diff, and apply share the same validator/materialization primitives, so their verdicts are identical by construction. The round-2 MEDIUM adjudication stands; no coordinator revisit needed.

### Notes on the updated pins
- The execute-after-rejected-approve assertion is kept in every test as `409 InvalidOperation` ("Cannot execute proposal in status PendingReview") — pinning that a statically invalid revision can no longer reach the apply stage at all, which is the trust-class goal.
- Board-state invariants (victim cards untouched, labels unchanged, no cross-board cards) are all retained.
- Tests whose names said `ShouldFailPreviewAndApply` but whose apply step is now unreachable were renamed `ShouldFailPreviewAndApprove`; the restructured ambiguity test keeps its `...PreviewAndApply` name because it now genuinely pins preview+apply post-approval.

### Verification (head `ae6892ec`, fix commit `ae6892ec`)
- Build clean, 0 errors.
- `ProposalRevisionApiTests`: **33 passed / 0 failed** (31 prior + 2 new static-ambiguity cases).
- `~AutomationProposalService` (Application): **86 passed / 0 failed**.
- Approve-touching Api regression set (ProposalErrorContract, AutomationProposalsApi, McpTools, ProposalApprovalRace, CaptureToBoardGoldenPath, CrossUserDataIsolation, CardsApi, CaptureApi, TranscriptTriageLlmGoldenPath, LiveBrowserRegressionApi, AuthzRegressionMatrix): **182 passed / 0 failed**.
- `ConcurrencyRaceConditionStressTests`: **30 passed / 0 failed**.

File: backend/src/Taskdeck.Application/Services/AutomationProposalService.cs [hunk 1]
```diff
@@ -234,6 +234,54 @@ public async Task<Result<ProposalDto>> ApproveProposalAsync(Guid id, Guid decide
...
+            // This closes the last "user commits to something the executor will refuse" step in
+            // this trust class (siblings #1370 → #1374, #1376 → #1395, #1398 → #1413).
+            //
+            // Gate only genuinely approvable (PendingReview) proposals: for any other status the
+            // domain transition's terminal-status short-circuit owns the response (409 "Cannot
+            // approve proposal in status X"), which this slice leaves untouched — running these
+            // gates on a terminal proposal would wrongly report a 400/403/404 in place of that 409.
+            if (proposal.Status == ProposalStatus.PendingReview)
+            {
+                var effectiveOperations = await ResolveEffectiveOperationsAsync(proposal, cancellationToken);
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/AutomationProposalService.cs:265]: Coordinator adjudication: **real, pre-existing, tracked as #1428** — not fixed in this PR.

The race is accurate: `CreateRevisionAsync` guards only on a `PendingReview` read and never touches the proposal row, so a revision-save that passes its guard before approve commits can insert a new latest revision after approve validated — and Apply's latest-revision materialization (which predates this PR) then executes content the reviewer never saw. Two mitigating facts scope it: Apply still re-runs the full gate chain transactionally, so an *invalid* racing revision is refused (no unsafe mutation — the safety boundary holds); and both the latest-revision materialization and the PendingReview-only guard predate this PR — #1416's approve==apply contract is what promotes the race from curiosity to first-class violation.

The fix is a revision-lifecycle design change (approved-revision pinning — the stronger semantics, which also intersects #1424's effective-operations question — or approve/revision mutual exclusion), out of this PR's scope. Tracked with full analysis and both design options in **#1428** per the review policy (no silent drops).
